### PR TITLE
Add per-module MFSDP prefetch budgets

### PR DESCRIPTION
**Pull request by wujingyue**
## Summary

- add independent forward and backward prefetch element budgets to MFSDP v2 `fully_shard()`
- use `suggested_communication_unit_size` as the adapter-level value passed to each wrapped module
- verify zero budgets disable all-gather/GEMM overlap

## Testing

- `python -m torch.distributed.run --nproc-per-node=2 -m pytest -q tests/unit_tests/distributed/mfsdp_v2/test_overlap.py::test_prefetch_size_zero_disables_allgather_overlap`
- `isort --check-only ... && black --check ... && ruff check ...`

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 1]
```diff
@@ -35,6 +35,13 @@ def _is_in_backward() -> bool:
     return torch._C._current_graph_task_id() != -1
 
 
+def _validate_prefetch_size(prefetch_size: int | None, name: str) -> int | None:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py]: Inline this. optimize for readability not conciseness.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 2]
```diff
@@ -348,9 +365,24 @@ def pre_forward(self) -> None:
         # prefetch the next module in forward order: its backward may already
         # be complete, so no later backward hook would reshard it.
         if not is_recomputing:
-            next_module = context.forward_order.next_item(self)
-            if next_module is not None:
-                next_module._unshard_parameter_groups()
+            self._prefetch_parameter_groups(context.forward_order, self._forward_prefetch_size)
+
+    def _prefetch_parameter_groups(
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:361]: Try your best to simplify

File: tests/unit_tests/distributed/mfsdp_v2/test_overlap.py [hunk 3]
```diff
@@ -251,3 +251,61 @@ def train_one_step() -> None:
 
     # Release the dedicated communicator so it does not leak into the shared session.
     dist.destroy_process_group(dp_group)
+
+
+@pytest.mark.flaky
+@pytest.mark.flaky_in_dev
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_overlap.py]: Why? Disabling overlap should be stable.

File: tests/unit_tests/distributed/mfsdp_v2/test_overlap.py [hunk 4]
```diff
@@ -251,3 +251,61 @@ def train_one_step() -> None:
...
+    device = distributed_setup.device
+    if world_size < 2:
+        pytest.skip("This test requires at least 2 ranks.")
+
+    dim = 16384
+    num_children = 4
+    dtype = torch.bfloat16
+    if not dist.is_initialized():
+        dist.init_process_group(backend="nccl")
+    dp_group = dist.new_group(backend="nccl")
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_overlap.py]: Why a new group? The default group doesn't work?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py (hunk 1 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py]: Done — the validation now sits directly in FsdpModule.__init__.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py (hunk 2 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py:361]: Simplified the control flow while keeping the shared forward/backward implementation.

File: tests/unit_tests/distributed/mfsdp_v2/test_overlap.py (hunk 3 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_overlap.py]: Removed both flaky markers. The focused two-rank profiler test passes.

File: tests/unit_tests/distributed/mfsdp_v2/test_overlap.py (hunk 4 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_overlap.py]: The default process group works. Switched the test to dist.group.WORLD and removed the dedicated group.

Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_overlap.py]: Agreed. This now uses init_device_mesh(device.type, (world_size,)); only the zero-CTA test needs a custom process group.

File: tests/unit_tests/distributed/mfsdp_v2/test_overlap.py [hunk 5]
```diff
@@ -251,3 +251,56 @@ def train_one_step() -> None:
...
+    def train_one_step() -> None:
+        optimizer.zero_grad(set_to_none=True)
+        for microbatch_index in range(2):
+            with microbatch(context, is_last=microbatch_index == 1):
+                model(x).sum().backward()
+        optimizer.step()
+
+    train_one_step()
+    torch.cuda.synchronize(device)
+    with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA]) as prof:
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_overlap.py:296]: Does profile do any synchronization already? It feels cumbersome to have to manually synchornize before and after profiling.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py [hunk 6]
```diff
@@ -108,6 +108,8 @@ def fully_shard(
     placements: Placements,
     mixed_precision_policy: MixedPrecisionPolicy | None = None,
     grad_divisor: int = 1,
+    forward_prefetch_size: int | None = None,
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: Do you recommend this or something like `fully_shard(...).set_forward_prefetch_size(size)`? I'm not sure.

File: tests/unit_tests/distributed/mfsdp_v2/test_overlap.py (hunk 5 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_overlap.py:296]: No. In this PyTorch version, profile.__exit__ only stops the profiler; it does not call torch.cuda.synchronize(). The pre-profile synchronization excludes warmup work, and the synchronization inside the context ensures queued CUDA events are recorded before profiling stops.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py (hunk 6 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: I recommend keeping keyword arguments on fully_shard(). They match the existing in-place construction API and configure the module before its hooks run. A chained setter would require changing fully_shard() from its current None-returning contract; a non-chained module setter adds a second configuration surface without a scheduling need after construction.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py]: Implemented this as SchedulePolicy. fully_shard() now takes one schedule_policy argument, which groups forward/backward prefetch budgets and leaves room for future scheduling controls without growing the function signature.

**Review by shjwudp — APPROVED** [review]

**Comment by wujingyue** [comment]
Hey @rapatel, let me know if you want this merged! Otherwise, I'll hold off to avoid unnecessary complexity.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/schedule.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+@dataclass(frozen=True)
+class SchedulePolicy:
+    """Control communication scheduling for one FSDP module.
+
+    ``None`` prefetches one successor, preserving the default behavior. ``0``
+    disables prefetching. Positive values specify parameter-element budgets.
+    """
+
+    forward_prefetch_size: int | None = None
```
Comment (rapatel) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/schedule.py:28]: So a positive value represents the number of bytes? Or something else?

Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/schedule.py:28]: Oh I see it's set by `ddp_config.suggested_communication_unit_size`

**Comment by rapatel** [comment]
@wujingyue would you mind holding off on merging this until my PR #6594 is merged?

**Comment by wujingyue** [comment]
> @wujingyue would you mind holding off on merging this until my PR #6594 is merged?

Sure. Just worried about merge conflicts?

**Comment by rapatel** [comment]
@wujingyue Yeah, just want to avoid the conflicts (I am not sure if there will be any though).

**Review by rapatel — APPROVED** [review]
LGTM

**Comment by rapatel** [comment]
BTW I noticed in `megatron_fsdp/megatron_fsdp.py`, the logic sets `suggested_communication_unit_size` if it is `none`. And then in the branch, it sets:
```
# Cap to 1B elements.
suggested_communication_unit_size = max(
    1_000_000_000, suggested_communication_unit_size
)
```

I think this should probably be a `min`.
